tools/github-actions/audit/src/reports.ts

```diff
@@ -0,0 +1,188 @@
+import * as core from '@actions/core';
+import type {
+  GitHubAdvisoryId,
+  NPMAuditReportV1,
+  NPMAuditReportV2,
+  Severity,
+  YarnAudit,
+  YarnNpmAuditReport
+} from 'audit-types';
+type Audit = NPMAuditReportV1.Audit;
+type Advisory = NPMAuditReportV1.Advisory;
+
+/**
+ * Severities supported by npm audit from the lowest to the highest criticality
+ */
+export const severities = ['info', 'low', 'moderate', 'high', 'critical'] as const satisfies Severity[];
+
+/**
+ * Description of an advisory as it will be displayed in the action summary
+ */
+export interface OtterAdvisory {
+  /** Severity of the vulnerability */
+  severity: Severity;
+  /** Overview of the vulnerability */
+  overview: string;
+  /** Name of the module affected by the vulnerability */
+  moduleName: string;
+}
+
+/**
+ * Data to build the action report summary
+ */
+export interface OtterAuditReport {
+  /** Total number of vulnerabilities */
+  nbVulnerabilities: number;
+  /** List of vulnerabilities that exceed the severity threshold */
+  errors: OtterAdvisory[];
+  /** List of vulnerabilities that don't exceed the severity threshold */
+  warnings: OtterAdvisory[];
+  /** Highest severity found in the list of vulnerabilities */
+  highestSeverityFound?: Severity;
+}
+
+/**
+ * Method which formats the response of an audit api (yarn or npm) into a report that can be displayed in the
+ * GitHub action summary
+ * @param response The output of the audit command
+ * @param severityThreshold The severity threshold to consider a vulnerability as an error
+ */
+export type OtterAuditReporter = (response: string, severityThreshold: Severity) => OtterAuditReport;
+
+function computeTotalsForSeverities(severityAccessor: (severity: Severity) => number) {
+  let highestSeverityFound: Severity | undefined;
+  let nbVulnerabilities = 0;
+  for (let index = severities.length - 1; index >= 0; index--) {
+    const severity: Severity = severities[index];
+    const severityTotal = severityAccessor(severity);
+    if (!highestSeverityFound && severityTotal > 0) {
+      highestSeverityFound = severity;
+    }
+    nbVulnerabilities += severityTotal;
+  }
+  return { highestSeverityFound, nbVulnerabilities };
+}
+
+function updateReportWithVulnerability(currentReport: OtterAuditReport, severityThreshold: Severity, vulnerability: OtterAdvisory): OtterAuditReport {
```

Comment:
it is only needed for Yarn4 because it doesn't provide a summary
The loop to find the highest severity actually loops only over 5 elements, it's not the same loop as the one on the list of vulnerabilities
So no, I don't think there is any reason not to trust the summary and compute the highest severity on our own